controllers/graph.go

```diff
@@ -0,0 +1,704 @@
+func (c *ApiController) UpdateGraph() {
+	id := c.Input().Get("id")
+	if id == "" {
+		c.ResponseError("id is empty")
+		return
+	}
+
+	var graph object.Graph
+	err := json.Unmarshal(c.Ctx.Input.RequestBody, &graph)
+	if err != nil {
+		c.ResponseError(err.Error())
+		return
+	}
+
+	graph.UpdatedTime = util.GetCurrentTime()
+
+	success, err := object.UpdateGraph(id, &graph)
+	if err != nil {
+		c.ResponseError(err.Error())
+		return
+	}
+
+	c.ResponseOk(success)
+}
+
+// AddGraph
+// @Title AddGraph
+// @Tag Graph API
+// @Description add graph
+// @Success 200 {object} controllers.Response The Response object
+// @router /add-graph [post]
+func (c *ApiController) AddGraph() {
+	var graph object.Graph
+	err := json.Unmarshal(c.Ctx.Input.RequestBody, &graph)
+	if err != nil {
+		c.ResponseError(err.Error())
+		return
+	}
+
+	owner := c.GetSessionUsername()
+	if owner == "" {
+		c.ResponseError("not signed in")
+		return
+	}
+
+	graph.Owner = owner
+	graph.CreatedTime = util.GetCurrentTime()
+	graph.UpdatedTime = util.GetCurrentTime()
+
+	if graph.GraphType == "" {
+		graph.GraphType = "force"
+	}
+
+	success, err := object.AddGraph(&graph)
+	if err != nil {
+		c.ResponseError(err.Error())
+		return
+	}
+
+	c.ResponseOk(success)
+}
+
+// DeleteGraph
+// @Title DeleteGraph
+// @Tag Graph API
+// @Description delete graph
+// @Success 200 {object} controllers.Response The Response object
+// @router /delete-graph [post]
+func (c *ApiController) DeleteGraph() {
+	var graph object.Graph
+	err := json.Unmarshal(c.Ctx.Input.RequestBody, &graph)
+	if err != nil {
+		c.ResponseError(err.Error())
+		return
+	}
+
+	success, err := object.DeleteGraph(&graph)
+	if err != nil {
+		c.ResponseError(err.Error())
+		return
+	}
+
+	c.ResponseOk(success)
+}
+
+// GenerateGraph
+// @Title GenerateGraph
+// @Tag Graph API
+// @Description generate graph data and analysis
+// @Success 200 {object} controllers.Response The Response object
+// @router /generate-graph [get]
+func (c *ApiController) GenerateGraph() {
+	id := c.Input().Get("id")
+	if id == "" {
+		c.ResponseError("id is empty")
+		return
+	}
+
+	graph, err := object.GetGraph(id)
+	if err != nil {
+		c.ResponseError(err.Error())
+		return
+	}
+
+	if graph == nil {
+		c.ResponseError("graph not found")
+		return
+	}
+
+	graphData, analysis, err := c.generateGraphData(graph)
+	if err != nil {
+		c.ResponseError(err.Error())
+		return
+	}
+
+	graphDataJson, _ := json.Marshal(graphData)
+	analysisJson, _ := json.Marshal(analysis)
+	graph.GraphData = string(graphDataJson)
+	graph.Analysis = string(analysisJson)
+	graph.UpdatedTime = util.GetCurrentTime()
+
+	success, err := object.UpdateGraph(id, graph)
+	if err != nil {
+		c.ResponseError(err.Error())
+		return
+	}
+
+	c.ResponseOk(map[string]interface{}{
+		"success":   success,
+		"graphData": graphData,
+		"analysis":  analysis,
+	})
+}
+
+func (c *ApiController) generateGraphData(graph *object.Graph) (object.GraphData, object.Analysis, error) {
```

Comment:
Put your new logic except CRUD (and has nothing to do with controller) into object package's graph_algorithm.go